zmon_aws_agent/elastigroup.py

```diff
@@ -0,0 +1,197 @@
+import logging
+import traceback
+
+import boto3
+import inflection
+from opentracing_utils import extract_span_from_kwargs, trace
+from spotinst_sdk import SpotinstClient
+
+from zmon_aws_agent.aws import entity_id, add_traffic_tags_to_entity, MAX_PAGE
+from zmon_aws_agent.common import call_and_retry
+
+ELASTIGROUP_RESOURCE_TYPE = 'Custom::elastigroup'
+
+logger = logging.getLogger(__name__)
+
+
+class Elastigroup:
+    """
+    Data required to access SpotInst API
+    """
+
+    def __init__(self, group_id, group_name, account_id, access_token):
+        self.group_id = group_id
+        self.group_name = group_name
+        self.account_id = account_id
+        self.access_token = access_token
+
+    def __eq__(self, other):
+        """Overrides the default implementation"""
+        if isinstance(other, Elastigroup):
+            return self.group_id == other.group_id and self.group_name == other.group_name and \
+                   self.account_id == other.account_id and self.access_token == other.access_token
+        return False
+
+
+@trace(tags={'aws': 'elastigroup'}, pass_span=True)
+def get_elastigroup_entities(region, acc, **kwargs):
+    groups = []
+    current_span = extract_span_from_kwargs(**kwargs)
+    current_span.set_tag("aws_region", region)
+    current_span.set_tag("account_id", acc)
+    cf = boto3.client('cloudformation', region_name=region)
+
+    stack_names = get_all_stack_names(cf)
+    for stack_name in stack_names:
+        elastigroups = get_elastigroup_resources(cf, stack_name)
+        for eg_data in elastigroups:
+            eg_details = get_elastigroup(eg_data, **kwargs)
+            eg = {
+                'id': entity_id('elastigroup-{}[{}:{}]'.format(eg_details['name'], acc, region)),
+                'type': 'elastigroup',
+                'infrastructure_account': acc,
+                'region': region,
+                'created_by': 'agent',
+                'name': eg_details['name'],
+                'availability_zones': [az['name'] for az in eg_details['compute']['availability_zones']],
+                'desired_capacity': eg_details['capacity']['target'],
+                'max_size': eg_details['capacity']['maximum'],
```

Comment:
Are the fields in `eg_details` guaranteed? otherwise we can be safe and use ``.get('..', None)``

Reply:
Fair question. According to the [respective API](https://api.spotinst.com/spotinst-api/elastigroup/amazon-web-services/list-group/ ), that's what I believed to expect.

Looking at the particular attribute `response.items.[].strategy.onDemandCount` and its value of `null` makes me believe that the attributes will be present, eventually with that kind of value.

I couldn't easily find in the Spotinst API docs which fields will always be present. My assumption, probably naive, is that they'll always will.

Reply:
Actually, changing to `get()` on our side wouldn't hurt and it becomes safer